src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+using OmniSharp.Extensions.JsonRpc;
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    internal abstract class AbstractRazorDelegatingEndpoint<TRequest, TResponse, TDelegatedParams> : IJsonRpcRequestHandler<TRequest, TResponse?>
+        where TRequest : TextDocumentPositionParams, IRequest<TResponse>
+        where TResponse : class
+        where TDelegatedParams : class
+    {
+        private readonly DocumentContextFactory _documentContextFactory;
+        private readonly LanguageServerFeatureOptions _languageServerFeatureOptions;
+        private readonly RazorDocumentMappingService _documentMappingService;
+        private readonly ClientNotifierServiceBase _languageServer;
+        private readonly ILogger<RazorHoverEndpoint> _logger;
+
+        protected AbstractRazorDelegatingEndpoint(
+            DocumentContextFactory documentContextFactory,
+            LanguageServerFeatureOptions languageServerFeatureOptions,
+            RazorDocumentMappingService documentMappingService,
+            ClientNotifierServiceBase languageServer,
+            ILoggerFactory loggerFactory)
+        {
+            _documentContextFactory = documentContextFactory ?? throw new ArgumentNullException(nameof(documentContextFactory));
+            _languageServerFeatureOptions = languageServerFeatureOptions ?? throw new ArgumentNullException(nameof(languageServerFeatureOptions));
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            _logger = loggerFactory?.CreateLogger<RazorHoverEndpoint>() ?? throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        /// <summary>
+        /// Override to provide fallback behavior for when single server is not supported
+        /// </summary>
+        protected abstract Task<TResponse?> HandleWithoutSingleServerAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Override to provide behavior for razor contexts. Since the c#/html servers won't
+        /// have any information for this context it doesn't need to be delegated
+        /// </summary>
+        protected abstract Task<TResponse?> HandleInRazorAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// The delegated object to send to the <see cref="EndpointName"/>
+        /// </summary>
+        protected abstract TDelegatedParams CreateDelegatedParams(TRequest request, DocumentContext documentContext, Projection projection);
```

Comment:
It'd be dope if we could eventually have a common base type / contract for delegated params which was understood by our server target on the client. For instance, if the delegated params had an "endpoint name" associated we could even remove the `EndpointName` bit below

Reply:
Originally I liked this comment, but now that I see what EndpointName is, I think that's a pretty leaky abstraction. it means the delegate params is the parameter object being delegated too, but also has information about how to delegate.

In fact, I think there wouldn't really be any benefit to having a base type for delegated params, as it wouldn't save anything. We'd still need the inheritor to construct it, just this method signature would look slightly different. A better goal would be to have a single target in Custom Message Target, which took a known parameter object, and inside that we had the delegated params. I know we talked about this before too, and I also think its not something for this PR.

Reply:
Want to think more about this, could be a good follow up PR from this after we've discussed a bit more